src/osd/scrubber/osd_scrub_sched.h

```diff
@@ -0,0 +1,349 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <atomic>
+#include <chrono>
+#include <memory>
+#include <optional>
+#include <vector>
+
+#include "common/RefCountedObj.h"
+#include "osd/osd_types.h"
+#include "osd/scrubber_common.h"
+
+#include "utime.h"
+
+class PG;
+
+namespace Scrub {
+
+// possible outcome when trying to select a PG and scrub it
+enum class attempt_t {
+  scrubbing,	    // successfully started a scrub
+  none_ready,	    // no pg to scrub
+  local_resources,  // failure to secure local OSD scrub resource
+  already_started,  // already started scrubbing this pg
+  no_pg,	    // can't find this pg
+  pg_state,	    // pg state (clean, active, etc.)
+  preconditions	    // time, configuration, etc.
+};
+
+}  // namespace Scrub
+
+/**
+ * the ordered queue of PGs waiting to be scrubbed.
+ * Main operations are scheduling/unscheduling a PG to be scrubbed at a certain
+ * time.
+ *
+ * A "penalty" queue maintains those PGs that have failed to reserve the
```

Comment:
IMO, "penalty" might be a little bit misleading. it's more like a backoff queue or a pending queue. as they are not punished for doing anything wrong like use more resources than they are entitled. they are just not lucky..